File: src/apps/Andriod/app/src/main/java/com/dkellycollins/triggerfinger/data/daos/impl/TouchPositionDao.java
package com.dkellycollins.triggerfinger.data.daos.impl;

import android.view.MotionEvent;
import android.view.View;

import com.dkellycollins.triggerfinger.data.daos.ITouchPositionDao;
import com.dkellycollins.triggerfinger.data.model.impl.Vector2;
import com.dkellycollins.triggerfinger.data.model.IVector;
import com.dkellycollins.triggerfinger.util.logger.ILogger;

public class TouchPositionDao implements ITouchPositionDao, View.OnTouchListener {

    private final View _view;
    private final Vector2 _position;

    private boolean _isTouching;

    public TouchPositionDao(View view) {
        _view = view;

        _position = new Vector2();
        _isTouching = false;

        _view.setOnTouchListener(this);
    }

    @Override
    public IVector getLastPosition() {
        return _position;
    }

    @Override
    public boolean isTouching() {
        return _isTouching;
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        _position.setX(event.getX());
        _position.setY(event.getY());

        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                _isTouching = true;
                break;
            case MotionEvent.ACTION_UP:
                _isTouching = false;
        }

        return true;
    }
}

File: src/apps/Android-Kotlin/app/src/main/kotlin/com/dkellycollins/triggerfinger/data/daos/impl/TouchPositionDao.kt
package com.dkellycollins.triggerfinger.data.daos.impl

import android.view.MotionEvent
import android.view.View

import com.dkellycollins.triggerfinger.data.daos.ITouchPositionDao
import com.dkellycollins.triggerfinger.data.model.impl.Vector2
import com.dkellycollins.triggerfinger.data.model.IVector
import com.dkellycollins.triggerfinger.util.logger.ILogger

class TouchPositionDao(private val _view: View) : ITouchPositionDao, View.OnTouchListener {
    private val _position: Vector2

    override var isTouching: Boolean = false
        private set(value: Boolean) {
            super.isTouching = value
        }

    init {

        _position = Vector2()
        isTouching = false

        _view.setOnTouchListener(this)
    }

    override val lastPosition: IVector
        get() = _position

    override fun onTouch(v: View, event: MotionEvent): Boolean {
        _position.x = event.x
        _position.y = event.y

        when (event.action) {
            MotionEvent.ACTION_DOWN -> isTouching = true
            MotionEvent.ACTION_UP -> isTouching = false
        }

        return true
    }
}
